hbase-server/src/main/resources/hbase-webapps/regionserver/rsNamedQueueLog.jsp

```diff
@@ -0,0 +1,157 @@
+<%--
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+--%>
+<%@ page contentType="text/html;charset=UTF-8"
+  import="java.util.Date"
+  import="java.util.List"
+  import="java.util.Arrays"
+  import="java.util.HashSet"
+  import="org.apache.hadoop.conf.Configuration"
+  import="org.apache.hadoop.hbase.client.Admin"
+  import="org.apache.hadoop.hbase.client.SnapshotDescription"
+  import="org.apache.hadoop.hbase.http.InfoServer"
+  import="org.apache.hadoop.hbase.master.HMaster"
+  import="org.apache.hadoop.hbase.snapshot.SnapshotInfo"
+  import="org.apache.hadoop.util.StringUtils"
+  import="org.apache.hadoop.hbase.TableName"
+  import="org.apache.hadoop.hbase.client.ServerType"
+  import="org.apache.hadoop.hbase.client.LogEntry"
+  import="org.apache.hadoop.hbase.regionserver.HRegionServer"
+  import="org.apache.hadoop.hbase.client.OnlineLogRecord"
+  import="org.apache.hadoop.hbase.HConstants"
+  import="org.apache.hadoop.hbase.ServerName"
+%>
+<%
+  HRegionServer rs = (HRegionServer) getServletContext().getAttribute(HRegionServer.REGIONSERVER);
+  Configuration conf = rs.getConfiguration();
+  List<OnlineLogRecord> slowLogs = null;
+  List<OnlineLogRecord> largeLogs = null;
+
+  if(rs.isOnline()) {
+    try (Admin rsAdmin = rs.getConnection().getAdmin()) {
+      slowLogs = (List<OnlineLogRecord>)(List<?>)rsAdmin.getLogEntries(new HashSet<ServerName>(Arrays.asList(rs.getServerName())), "SLOW_LOG", ServerType.REGION_SERVER, HConstants.DEFAULT_SLOW_LOG_RING_BUFFER_SIZE, null);
```

Comment:
So, in general, we need to get this information from master? Then why not provide it on master status site...

Reply:
@Apache9 Basically the ring buffers that store this slow/large RPC log info operate at RegionServer side, hence individual RegionServer will have it's own ring buffer to provide this logs. As opposed to say BalancerDecision or BalancerRejection logs (which are also built on the same framework - NamedQueue) have ring buffers running at master side, so they are displayed only at master page.

Reply:
If they are on region server, then why we need to use Admin to get these slow logs...

Reply:
I see your point. We have generic Admin API to retrieve all kind of log records, regardless of whether ring buffers operate on Master or RegionServer.

Admin#getLogEntries [here](https://github.com/apache/hbase/blob/master/hbase-client/src/main/java/org/apache/hadoop/hbase/client/Admin.java#L2532)

```
  /**
   * Retrieve recent online records from HMaster / RegionServers.
   * Examples include slow/large RPC logs, balancer decisions by master.
   *
   * @param serverNames servers to retrieve records from, useful in case of records maintained
   *   by RegionServer as we can select specific server. In case of servertype=MASTER, logs will
   *   only come from the currently active master.
   * @param logType string representing type of log records
   * @param serverType enum for server type: HMaster or RegionServer
   * @param limit put a limit to list of records that server should send in response
   * @param filterParams additional filter params
   * @return Log entries representing online records from servers
   * @throws IOException if a remote or network exception occurs
   */
  List<LogEntry> getLogEntries(Set<ServerName> serverNames, String logType,
    ServerType serverType, int limit, Map<String, Object> filterParams) throws IOException;

```

Reply:
Based on logType client connect master or RegionServer,
https://github.com/apache/hbase/blob/479ae88be380db46eb1d64c0808738b52bd9f605/hbase-client/src/main/java/org/apache/hadoop/hbase/client/RawAsyncHBaseAdmin.java#L4257

Here we are going to show slow/large logs of RegionSevrver, so instead of making Admin RPC we should directly fetch from HRegionServer.

Reply:
@Apache9 @pankaj72981 As @virajjasani  says, there is a generic admin API to retrieve all kinds of log entries. If getting logs from RS by admin API is not a good way. Let me think about any good way to retrieve it directly from RS

Reply:
@GeorryHuang Thanks, yes you can try to explore if you are able to directly fetch records from RegionServer.